apps/mobile/src/features/swap/swap-state/use-swap-state.ts

```diff
@@ -0,0 +1,202 @@
+    slippageEditingAllowed: true,
+    selectingAsset: null,
+    validation: [],
+  };
+}
+
+export interface UseSwapStateProps {
+  accountRequest: { account: AccountAddresses };
+  baseAsset?: FungibleCryptoAsset;
+  targetAsset?: FungibleCryptoAsset;
+  marketDataService: MarketDataService;
+  swapService: SwapService;
+  // TODO: Come up with a better name. This is an adapter for assetVisibility.
+  isAssetAllowed?: (asset: FungibleCryptoAsset) => boolean;
+  quoteCurrencyPreference: QuoteCurrency;
+}
+
+export function useSwapState({
+  accountRequest,
+  swapService,
+  marketDataService,
+  baseAsset,
+  targetAsset,
+  quoteCurrencyPreference,
+  isAssetAllowed,
+}: UseSwapStateProps): UseSwapStateResult {
+  const [state, dispatch] = useReducer(
+    swapReducer,
+    { baseAsset, targetAsset, quoteCurrencyPreference },
+    initializeState
+  );
+
+  const queries = useMemo(() => {
+    return {
+      useAccountBaseSwapAssetsQuery: createAccountBaseSwapAssetsQuery(swapService, accountRequest),
+      useAccountTargetSwapAssetsQuery: createAccountTargetSwapAssetsQuery(
+        swapService,
+        accountRequest
+      ),
+      useAssetMarketDataQuery: createAssetMarketDataQuery(marketDataService),
+    };
+  }, [accountRequest, swapService, marketDataService]);
+
+  const baseAssetsQuery = queries.useAccountBaseSwapAssetsQuery({
+    queryOptions: { select: createSwapAssetsSelector('base', isAssetAllowed) },
+  });
+  const targetAssetsQuery = queries.useAccountTargetSwapAssetsQuery({
+    baseId: state.baseSwapAsset ? getAssetId(state.baseSwapAsset?.asset) : undefined,
+    queryOptions: { select: createSwapAssetsSelector('target', isAssetAllowed) },
+  });
+  const marketDataQuery = queries.useAssetMarketDataQuery({
+    asset: state.baseSwapAsset?.asset ?? null,
+  });
+
+  const { derivedAmounts, lockDerivedAmountsForNextRender } = useDerivedAmounts(
+    state,
+    marketDataQuery.data
+  );
+
+  useEffect(() => {
+    if (baseAssetsQuery.data) {
+      dispatch({ type: 'RECONCILE_BASE_WITH_PROVIDER', payload: baseAssetsQuery.data });
+    }
+  }, [baseAssetsQuery.data]);
+
+  useEffect(() => {
+    if (targetAssetsQuery.data) {
+      dispatch({ type: 'RECONCILE_TARGET_WITH_PROVIDER', payload: targetAssetsQuery.data });
+    }
+  }, [targetAssetsQuery.data]);
+
+  const setBaseSwapAsset = useCallback((asset: AccountSwapAsset) => {
+    dispatch({ type: 'SET_BASE_SWAP_ASSET', payload: asset });
+  }, []);
+
+  const setTargetSwapAsset = useCallback((asset: AccountSwapAsset) => {
+    dispatch({ type: 'SET_TARGET_SWAP_ASSET', payload: asset });
+  }, []);
+
+  const clearAssetSelection = useCallback(() => {
+    dispatch({ type: 'CLEAR_ASSET_SELECTION' });
+  }, []);
+
+  const flipAssets = useCallback(() => {
+    dispatch({ type: 'FLIP_ASSETS' });
+  }, []);
+
+  const setBaseAmount = useCallback((amount: string) => {
+    dispatch({ type: 'SET_BASE_AMOUNT', payload: amount });
+  }, []);
```

Comment:
Is `useCallback` necessary here? These are such simple dispatch events.

Reply:
```ts
const actions = useMemo(() => ({
  setSlippage: (slippage: number) => {
    dispatch({ type: 'SET_SLIPPAGE', payload: slippage });
  },
  toggleSlippageEditing: (allowed: boolean) => {
    dispatch({ type: 'TOGGLE_SLIPPAGE_EDITING', payload: allowed });
  },
  setNonce: (nonce: number) => {
    dispatch({ type: 'SET_NONCE', payload: nonce });
  },
  openAssetSelector: (type: 'base' | 'target') => {
    dispatch({ type: 'OPEN_ASSET_SELECTOR', payload: type });
  },
  closeAssetSelector: () => {
    dispatch({ type: 'CLOSE_ASSET_SELECTOR' });
  },
}), [dispatch]);
```

a single memo would have same perf benefits with tidier syntax

Reply:
That's clever, didn't think of this. In my brain useMemo === data. Let me validate if we need this in the first place then refactor. Might do it in the next PR if profiling doesn't work out of the box.